### PR TITLE
Smarter field qualifying algorithm

### PR DESCRIPTION
**Pull request by gavv**
This PR implements smarter algorithm for field qualifier, discussed in #387.

New algorithm tries all struct type/field names, recursively for nested structs. Additional syntax for named extends (`xorm:"extends(name)"`) is **not** implemented yet.

Two related PRs:
* go-xorm/core#14
* go-xorm/tests#13 (much more extends tests for various tricky cases)

### CODE REVIEW

**Comment by lunny** [comment]
I think there are still many situation cannot be resolved. `statement.tableMap        map[string][]string` maybe should be `tableMap       [][]string`. And the sequence should be the same as `Join` sequence. So that it's easy to column to find there table name or alias name.

**Comment by gavv** [comment]
> `statement.tableMap map[string][]string` maybe should be `tableMap [][]string`

I don't quite understand how should we use it. Could you clarify?

Currently, this PR introduces the following behaviour:
* if table name can be deduced **unambiguously** for all fields, use qualified fields
* otherwise, if table name can't be deduced unambiguously at least for one field, emit warning and fallback to `SELECT *`

See [`isKnownUnambiguousTable()`](https://github.com/gavv/xorm/blob/facf54e755a562fdb98b246109e8cbed64a06101/statement.go#L471).

In other words, new behaviour (qualified fields) is enabled only when we're quite sure that this will not break old code that relied on ambiguous fields.

> I think there are still many situation cannot be resolved.

What are those situations? Let me know, I'll add tests for them :)
